### PR TITLE
Refactor UENV based configurations

### PR DESCRIPTION
**Pull request by teojgo**

### CODE REVIEW

File: config/cscs.py [hunk 1]
```diff
@@ -15,7 +15,7 @@
 sys.path.append(utilities_path)
 
 import firecrest_slurm
-
+import uenv
```
Comment (jgphpc) [inline config/cscs.py:18]:
```suggestion
import uenv

```

File: config/cscs.py [hunk 2]
```diff
@@ -24,8 +24,7 @@ def is_var_true(var):
     return var.lower() in ['true', 'yes', '1']
```
Comment (jgphpc) [inline config/cscs.py:25]:
```suggestion


```

File: config/cscs.py [hunk 3]
```diff
@@ -45,3 +44,21 @@ def is_var_true(var):
...
+# names valid for that system with the ones from uenv
+if site_configuration and uenv_environs:
+    site_configuration['environments'] += uenv_environs
+    for system in site_configuration['systems']:
+        valid_system_uenv_names = [
+            u['name'] for u in uenv_environs 
+            if system['name'] in u['target_systems']
+        ]
+        for partition in system['partitions']:
+            if (partition.get('features', None) and
```
Comment (jgphpc) [inline config/cscs.py:60]:
```suggestion
            if (partition.get('features', None) and
```

File: config/cscs.py [hunk 4]
```diff
@@ -45,3 +44,21 @@ def is_var_true(var):
...
+if site_configuration and uenv_environs:
+    site_configuration['environments'] += uenv_environs
+    for system in site_configuration['systems']:
+        valid_system_uenv_names = [
+            u['name'] for u in uenv_environs 
+            if system['name'] in u['target_systems']
+        ]
+        for partition in system['partitions']:
+            if (partition.get('features', None) and 
+                'uenv' in partition['features']):
```
Comment (jgphpc) [inline config/cscs.py:61]:
```suggestion
               'uenv' in partition['features']):
```

File: config/utilities/uenv.py [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+_UENV_CLI = 'uenv'
+_UENV_DELIMITER = ','
+_UENV_MOUNT_DELIMITER = '@'
+_RFM_META = pathlib.Path('extra') / 'reframe.yaml'
+
+
+def _get_uenvs():
+    uenv = os.environ.get('UENV', None)
+    if uenv is None:
+        return uenv
```
Comment (jgphpc) [inline config/utilities/uenv.py:20]:
```suggestion
        return uenv
```

File: config/utilities/uenv.py [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+
+        image_path = osext.run_command(f"{inspect_cmd} '{{sqfs}}'", shell=True
+            ).stdout.strip()
+        target_system = osext.run_command(f"{inspect_cmd} '{{system}}'",
+            shell=True).stdout.strip()
+
+        image_path = pathlib.Path(image_path)
+
+        #FIXME temporary workaround for older uenv versions
+        if Version(uenv_version) >= Version('5.1.0-dev'):
```
Comment (jgphpc) [inline config/utilities/uenv.py:45]:
```suggestion
        if Version(uenv_version) > Version('5.1.0-dev'):
```

File: config/utilities/uenv.py [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+                f"{inspect_cmd} '{{meta}}'", shell=True
+            ).stdout.strip()
+            rfm_meta = pathlib.Path(meta_path) / _RFM_META
+        else:
+            rfm_meta = image_path.parent / 'store.yaml'
+
+        try:
+            with open(rfm_meta) as image_envs:
+                image_environments = yaml.load(
+                    image_envs.read(), Loader=yaml.BaseLoader)
```
Comment (jgphpc) [inline config/utilities/uenv.py:56]:
```suggestion
                    image_envs.read(), Loader=yaml.BaseLoader)
                    print(f"# --- loading the metadata from '{rfm_meta}'")
```

File: config/utilities/uenv.py [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+        except OSError as err:
+            raise ConfigError(
+                f"problem loading the metadata from '{rfm_meta}'"
+            )
+    
+        for k, v in image_environments.items():
+            env = {
+                'target_systems': [target_system]
```
Comment (jgphpc) [inline config/utilities/uenv.py:64]:
```suggestion
                'target_systems': [target_system]
```

File: config/utilities/uenv.py [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+            )
+    
+        for k, v in image_environments.items():
+            env = {
+                'target_systems': [target_system] 
+            }
+            env.update(v)
+    
+            activation = v['activation']
+
```
Comment (jgphpc) [inline config/utilities/uenv.py:69]:
```suggestion

```

File: config/utilities/uenv.py [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+            activation = v['activation']
+            
+            # FIXME: Assume that an activation script is given, to be sourced
+            if isinstance(activation, str):
+                if not activation.startswith(image_mount):
+                    raise ConfigError(
+                        f'activation script of {k!r} is not consistent '
+                        f'with the mount point: {image_mount!r}'
+                    )
+
```
Comment (jgphpc) [inline config/utilities/uenv.py:77]:
```suggestion

```

File: config/utilities/uenv.py [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+            uenv_name_pretty = uenv_name.replace(":", "_").replace("/", "_")
+            env['name'] = f'{uenv_name_pretty}_{k}'
+
+            env['resources'] = {
+                'uenv': {
+                    'mount': image_mount,
+                    'file': str(image_path),
+                }
+            }
+
```
Comment (jgphpc) [inline config/utilities/uenv.py:98]:
```suggestion

```

File: config/utilities/uenv.py [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+            env['resources'] = {
+                'uenv': {
+                    'mount': image_mount,
+                    'file': str(image_path),
+                }
+            }
+    
+            # Added as a prepare_cmd for the environment
+            del env['activation']
+
```
Comment (jgphpc) [inline config/utilities/uenv.py:101]:
```suggestion

```

**Review by jgphpc — APPROVED** [review]

File: config/utilities/uenv.py (hunk 7 above)
Comment (teojgo) [reply config/utilities/uenv.py:56]: I don't want to add this because it messes up the `--show-config` if I want to parse with `jq`

**Pull request merged**